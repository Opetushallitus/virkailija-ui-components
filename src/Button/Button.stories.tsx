import * as React from 'react';
import { storiesOf } from '@storybook/react';
import Button from './index';
import { select, boolean } from '@storybook/addon-knobs';

const stories = storiesOf('Components/Button', module);

stories.add('Default', () => {
  const variant = select(
    'Variant',
    {
      Contained: 'contained',
      Outlined: 'outlined',
      Text: 'text',
    },
    'contained',
  ) as any;

  const color = select(
    'Color',
    {
      Primary: 'primary',
      Secondary: 'secondary',
      Success: 'success',
      Danger: 'danger',
    },
    'primary',
  ) as any;

  const size = select(
    'Size',
    {
      Small: 'small',
      Medium: 'medium',
    },
    'medium',
  ) as any;

  const disabled = boolean('Disabled', false) as boolean;

  const fullWidth = boolean('Full width', false) as boolean;

  return (
    <Button
      variant={variant}
      color={color}
      size={size}
      disabled={disabled}
      fullWidth={fullWidth}
    >
      Button
    </Button>
  );
});
