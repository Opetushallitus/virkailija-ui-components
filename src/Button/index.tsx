import * as React from 'react';
import styled, { css, DefaultTheme } from 'styled-components';
import { disabledStyle } from '../system';
import HtmlButton from '../HtmlButton';

type ButtonVariant = 'contained' | 'outlined' | 'text';

type ButtonSize = 'small' | 'medium';

type ButtonColor = 'primary' | 'success' | 'danger' | 'secondary';

const paddings = {
  contained: {
    medium: '6px 16px',
    small: '4px 10px',
  },
  outlined: {
    medium: '4px 16px',
    small: '2px 10px',
  },
  text: {
    medium: '6px 8px',
    small: '4px 5px',
  },
} as const;

const getOutlinedColorStyle = ({
  color,
  theme,
}: {
  color: ButtonColor;
  theme: DefaultTheme;
}) => {
  const {
    outlineColor,
    hoverOutlineColor,
    focusOutlineColor,
  } = theme.buttonVariants.outlined[color];

  return `
    padding: 4px 16px;
    border: 2px solid ${outlineColor};
    color: ${outlineColor};

    &:hover, &:active {
      border-color: ${hoverOutlineColor};
      color: ${hoverOutlineColor};
    }

    &:focus {
      box-shadow: 0 0 0 3px ${focusOutlineColor};
    }
  `;
};

const getContainedColorStyle = ({
  color,
  theme,
}: {
  color: ButtonColor;
  theme: DefaultTheme;
}) => {
  const {
    backgroundColor,
    color: fontColor,
    hoverBackgroundColor,
    focusOutlineColor,
  } = theme.buttonVariants.contained[color];

  return `
    border-color: ${backgroundColor};
    background-color: ${backgroundColor};
    color: ${fontColor};

    &:hover, &:active {
      border-color: ${hoverBackgroundColor};
      background-color: ${hoverBackgroundColor};
    }

    &:focus {
      box-shadow: 0 0 0 3px ${focusOutlineColor};
    }
  `;
};

const getTextVariantColorStyle = ({
  color,
  theme,
}: {
  color: ButtonColor;
  theme: DefaultTheme;
}) => {
  const {
    color: fontColor,
    hoverColor,
    focusOutlineColor,
  } = theme.buttonVariants.text[color];

  return `
    border-color: transparent;
    background-color: transparent;
    color: ${fontColor};

    &:hover, &:active {
      color: ${hoverColor}
    }

    &:focus {
      box-shadow: 0 0 0 3px ${focusOutlineColor};
    }
  `;
};

const getVariantStyle = ({ variant }: { variant: ButtonVariant }) => {
  if (variant === 'outlined') {
    return css`
      background-color: transparent;
      ${getOutlinedColorStyle}
    `;
  } else if (variant === 'contained') {
    return css`
      ${getContainedColorStyle}
    `;
  } else if (variant === 'text') {
    return css`
      ${getTextVariantColorStyle}
      padding: 6px 8px;
    `;
  }
};

const getSizeStyle = ({
  size,
  variant,
  theme,
}: {
  size: ButtonSize;
  variant: ButtonVariant;
  theme: DefaultTheme;
}) => {
  return {
    fontSize: size === 'small' ? '0.85rem' : theme.fontSizes.body,
    padding: paddings[variant][size],
  };
};

const ButtonBase = styled.button<{
  size: ButtonSize;
  variant: ButtonVariant;
  disabled: boolean;
  color: ButtonColor;
  fullWidth: boolean;
}>`
  cursor: pointer;
  border: 0px none;
  box-shadow: 0 0 0 0 transparent;
  outline: none;
  padding: 6px 16px;
  border-radius: ${({ theme }) => theme.radii[1]}px;
  font-family: ${({ theme }) => theme.fonts.main};
  font-size: ${({ theme }) => theme.fontSizes.body};
  line-height: ${({ theme }) => theme.lineHeights.body};
  white-space: nowrap;
  display: inline-flex;
  align-items: center;
  box-sizing: border-box;
  transition: box-shadow 0.25s, background-color 0.25s, border-color 0.25s, color 0.25s;
  font-weight: ${({ theme }) => theme.fontWeights.bold};
  text-decoration: none;

  ${getVariantStyle}
  ${getSizeStyle}
  ${disabledStyle}

  ${({ fullWidth }) =>
    fullWidth &&
    css`
      width: 100%;
      justify-content: center;
    `}
`;

type ButtonBaseProps = {
  size?: ButtonSize;
  variant?: ButtonVariant;
  color?: ButtonColor;
  disabled?: boolean;
  fullWidth?: boolean;
  children: React.ReactNode;
};

export type ButtonProps = ButtonBaseProps &
  Omit<React.ComponentProps<typeof ButtonBase>, keyof ButtonBaseProps>;

const Button = React.forwardRef<HTMLElement, ButtonProps>(
  (
    {
      size = 'medium',
      variant = 'contained',
      color = 'primary',
      as: asProp = HtmlButton,
      disabled = false,
      fullWidth = false,
      ...props
    },
    ref,
  ) => {
    return (
      <ButtonBase
        ref={ref as React.Ref<HTMLButtonElement>}
        size={size}
        variant={variant}
        disabled={disabled}
        color={color}
        as={asProp}
        fullWidth={fullWidth}
        {...props}
      />
    );
  },
);

export default Button;
