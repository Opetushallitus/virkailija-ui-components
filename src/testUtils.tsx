import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { createTheme } from './createTheme';

const theme = createTheme();

const Providers = ({ children }: { children: React.ReactNode }) => (
  <ThemeProvider theme={theme}>{children}</ThemeProvider>
);

const dummyDiv = document.createElement('div');
dummyDiv.setAttribute('data-nothing', 'true');

// Return rootElement from render() for easier snapshot testing
const customRender = (ui: React.ReactElement, options: any = {}) => {
  const result = render(ui, { wrapper: Providers, ...options });
  return {
    ...result,
    rootElement: result.container.firstChild || dummyDiv,
  };
};

export * from '@testing-library/react';

export { customRender as render };
