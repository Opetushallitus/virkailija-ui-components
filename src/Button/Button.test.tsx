import React from 'react';

import Button from './index';
import { render, fireEvent } from '../testUtils';

test('Renders correctly with defaults', () => {
  const { rootElement } = render(<Button>Default button</Button>);
  expect(rootElement).toMatchSnapshot();
});

test('Renders correctly with custom size, variant, and color', () => {
  const { rootElement } = render(
    <Button size="small" variant="outlined" color="secondary">
      Small outlined secondary button
    </Button>,
  );
  expect(rootElement).toMatchSnapshot();
});

test('Renders correctly with custom component ("as"-prop)', () => {
  const CustomButton = (props = {}) => <div {...props} />;
  const { rootElement } = render(
    <Button as={CustomButton}>Button with custom component</Button>,
  );
  expect(rootElement).toMatchSnapshot();
});

test('Calls onClick when clicked', () => {
  const handleClick = jest.fn();
  const { rootElement } = render(
    <Button onClick={handleClick}>Click me!</Button>,
  );

  fireEvent.click(rootElement);
  expect(handleClick).toBeCalledTimes(1);
  expect(rootElement).toMatchSnapshot();
});

test('Disabled button cannot be interacted with', () => {
  const handleClick = jest.fn();
  const { rootElement } = render(
    <Button onClick={handleClick} disabled={true}>
      Click me!
    </Button>,
  );

  expect(handleClick).not.toBeCalled();
  expect(rootElement).toMatchSnapshot();
});

test('Loading state button cannot be interacted with', () => {
  const handleClick = jest.fn();
  const { rootElement } = render(
    <Button onClick={handleClick} loading={true}>
      Click me!
    </Button>,
  );
  const children = rootElement.childNodes;
  expect(handleClick).not.toBeCalled();
  expect(rootElement).toMatchSnapshot();
});

test('Loading state button displays Oph Spinner', () => {
  const { rootElement } = render(<Button loading={true}>Click me!</Button>);
  const children = rootElement.childNodes;
  expect((children[2] as HTMLDivElement).getAttribute('class')).toContain(
    'Oph-Spinner',
  );
  expect(rootElement).toMatchSnapshot();
});
