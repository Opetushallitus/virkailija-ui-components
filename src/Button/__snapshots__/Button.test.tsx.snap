// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calls onClick when clicked 1`] = `
.c0 {
  cursor: pointer;
  border: 0px none;
  box-shadow: 0 0 0 0 transparent;
  outline: none;
  padding: 6px 16px;
  border-radius: 4px;
  font-family: 'Roboto',sans-serif;
  font-size: 1rem;
  line-height: 1.5;
  white-space: nowrap;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
  -webkit-transition: box-shadow 0.25s,background-color 0.25s,border-color 0.25s, color 0.25s;
  transition: box-shadow 0.25s,background-color 0.25s,border-color 0.25s, color 0.25s;
  font-weight: 500;
  -webkit-text-decoration: none;
  text-decoration: none;
  border-color: #0a789c;
  background-color: #0a789c;
  color: #ffffff;
  font-size: 1rem;
  padding: 6px 16px;
}

.c0:hover,
.c0:active {
  border-color: #0d9dcc;
  background-color: #0d9dcc;
}

.c0:focus {
  box-shadow: 0 0 0 3px rgba(10,120,156,0.2);
}

<button
  class="c0"
  color="primary"
  type="button"
  variant="contained"
>
  Click me!
   
</button>
`;

exports[`Disabled button cannot be interacted with 1`] = `
.c0 {
  cursor: pointer;
  border: 0px none;
  box-shadow: 0 0 0 0 transparent;
  outline: none;
  padding: 6px 16px;
  border-radius: 4px;
  font-family: 'Roboto',sans-serif;
  font-size: 1rem;
  line-height: 1.5;
  white-space: nowrap;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
  -webkit-transition: box-shadow 0.25s,background-color 0.25s,border-color 0.25s, color 0.25s;
  transition: box-shadow 0.25s,background-color 0.25s,border-color 0.25s, color 0.25s;
  font-weight: 500;
  -webkit-text-decoration: none;
  text-decoration: none;
  border-color: #0a789c;
  background-color: #0a789c;
  color: #ffffff;
  font-size: 1rem;
  padding: 6px 16px;
  cursor: not-allowed;
  opacity: 0.5;
}

.c0:hover,
.c0:active {
  border-color: #0d9dcc;
  background-color: #0d9dcc;
}

.c0:focus {
  box-shadow: 0 0 0 3px rgba(10,120,156,0.2);
}

<button
  class="c0"
  color="primary"
  disabled=""
  type="button"
  variant="contained"
>
  Click me!
   
</button>
`;

exports[`Renders correctly with custom component ("as"-prop) 1`] = `
.c0 {
  cursor: pointer;
  border: 0px none;
  box-shadow: 0 0 0 0 transparent;
  outline: none;
  padding: 6px 16px;
  border-radius: 4px;
  font-family: 'Roboto',sans-serif;
  font-size: 1rem;
  line-height: 1.5;
  white-space: nowrap;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
  -webkit-transition: box-shadow 0.25s,background-color 0.25s,border-color 0.25s, color 0.25s;
  transition: box-shadow 0.25s,background-color 0.25s,border-color 0.25s, color 0.25s;
  font-weight: 500;
  -webkit-text-decoration: none;
  text-decoration: none;
  border-color: #0a789c;
  background-color: #0a789c;
  color: #ffffff;
  font-size: 1rem;
  padding: 6px 16px;
}

.c0:hover,
.c0:active {
  border-color: #0d9dcc;
  background-color: #0d9dcc;
}

.c0:focus {
  box-shadow: 0 0 0 3px rgba(10,120,156,0.2);
}

<div
  class="c0"
  color="primary"
  variant="contained"
>
  Button with custom component
   
</div>
`;

exports[`Renders correctly with custom size, variant, and color 1`] = `
.c0 {
  cursor: pointer;
  border: 0px none;
  box-shadow: 0 0 0 0 transparent;
  outline: none;
  padding: 6px 16px;
  border-radius: 4px;
  font-family: 'Roboto',sans-serif;
  font-size: 1rem;
  line-height: 1.5;
  white-space: nowrap;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
  -webkit-transition: box-shadow 0.25s,background-color 0.25s,border-color 0.25s, color 0.25s;
  transition: box-shadow 0.25s,background-color 0.25s,border-color 0.25s, color 0.25s;
  font-weight: 500;
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: transparent;
  padding: 4px 16px;
  border: 2px solid #666;
  color: #666;
  font-size: 0.85rem;
  padding: 2px 10px;
}

.c0:hover,
.c0:active {
  border-color: #808080;
  color: #808080;
}

.c0:focus {
  box-shadow: 0 0 0 3px rgba(102,102,102,0.2);
}

<button
  class="c0"
  color="secondary"
  type="button"
  variant="outlined"
>
  Small outlined secondary button
   
</button>
`;

exports[`Renders correctly with defaults 1`] = `
.c0 {
  cursor: pointer;
  border: 0px none;
  box-shadow: 0 0 0 0 transparent;
  outline: none;
  padding: 6px 16px;
  border-radius: 4px;
  font-family: 'Roboto',sans-serif;
  font-size: 1rem;
  line-height: 1.5;
  white-space: nowrap;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
  -webkit-transition: box-shadow 0.25s,background-color 0.25s,border-color 0.25s, color 0.25s;
  transition: box-shadow 0.25s,background-color 0.25s,border-color 0.25s, color 0.25s;
  font-weight: 500;
  -webkit-text-decoration: none;
  text-decoration: none;
  border-color: #0a789c;
  background-color: #0a789c;
  color: #ffffff;
  font-size: 1rem;
  padding: 6px 16px;
}

.c0:hover,
.c0:active {
  border-color: #0d9dcc;
  background-color: #0d9dcc;
}

.c0:focus {
  box-shadow: 0 0 0 3px rgba(10,120,156,0.2);
}

<button
  class="c0"
  color="primary"
  type="button"
  variant="contained"
>
  Default button
   
</button>
`;
